### PR TITLE
Assignment1

### PR DESCRIPTION
**Pull request by Bhavishya-T**

### CODE REVIEW

File: Assignment1/assignment1/lib/Types.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class Type {
+  double calculateTax(double price) {
+    return 0.0;
+  }
+
+  double getTotal(double tax, double price, int quanitity) {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: No need to pass the tax here, you can directly call the calculateTax method here and get it.

File: Assignment1/assignment1/lib/Types.dart [hunk 2]
```diff
@@ -0,0 +1,38 @@
+class Type {
+  double calculateTax(double price) {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: The class can be made abstract and default implementation can be skipped. An abstract class can have both implemented and un-implemented methods.

Comment (bhavyadeep-purswani) [reply Assignment1/assignment1/lib/Types.dart]: Also instead of passing the price in all the methods, the price can be made a data member of the class and accepted as part of the constructor. Same goes for quantity

File: Assignment1/assignment1/lib/assignment1.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import 'types.dart' as types;
+
+void generateBill(String name, double price, int quantity, String itemType) {
+  double tax = 0.0;
+  double total = 0.0;
+  itemType.toLowerCase();
+  switch (itemType) {
+    case 'raw':
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/assignment1.dart]: Instead of directly using string literal here create constant variables and use that.

File: Assignment1/assignment1/test/assignment1_test.dart [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    test("Item 3 - Imported", () {
+      var imported = types.Imported();
+      expect(1294.8, imported.getTotal(13.6, 86, 13));
+    });
+  });
+  group("Testing exception handling", () {
+    test("isPositive()", () {
+      expect(() => handler.isPositive(-10), throwsException);
+    });
+  });
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/test/assignment1_test.dart:42]: Add a positive test case also for isPositive()

File: Assignment1/assignment1/lib/Types.dart [hunk 5]
```diff
@@ -0,0 +1,38 @@
+class Type {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: Can be renamed to Item so that it's more descriptive.

File: Assignment1/assignment1/lib/Types.dart [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  double calculateTax(double price) {
+    return 0.0;
+  }
+
+  double getTotal(double tax, double price, int quanitity) {
+    return (tax + price) * quanitity;
+  }
+}
+
+class Raw extends Type {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: Can be renamed to RawItem so that it's more descriptive.

File: Assignment1/assignment1/lib/Types.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  @override
+  double calculateTax(double price) {
+    return price * 0.125;
+  }
+}
+
+class Manufactured extends Type {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: Can be renamed to ManufacturedItem so that it's more descriptive.

File: Assignment1/assignment1/lib/Types.dart [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  @override
+  double calculateTax(double price) {
+    return price * 0.125 + (price * 0.125 + price) * 0.02;
+  }
+}
+
+class Imported extends Type {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: Can be renamed to ImportedItem so that it's more descriptive.

File: Assignment1/assignment1/lib/handler.dart [hunk 9]
```diff
@@ -0,0 +1,6 @@
+void isPositive(var number) {
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/handler.dart]: 1. Can be wrapped in a class name Utility or UtitlityFunctions and the method can be made static.
2. Can be renamed to `ensurePositive` as `isPositive` gives an indication that it would return a boolean but in this case, it doesn't.
3. Instead of using var as an argument type, we can specify double

File: Assignment1/assignment1/lib/Types.dart [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  @override
+  double calculateTax(double price) {
+    double tax = price * 0.1;
```
Comment (bhavyadeep-purswani) [inline Assignment1/assignment1/lib/Types.dart]: Extract the constant values like 0.1, 5, 10 etc. into constants and name them appropriately so that it is easier to understand what a particular value represents. For e.g., 0.1 can be named as IMPORT_DUTY_PERCENTAGE.

File: Assignment1/assignment1/test/assignment1_test.dart (hunk 4 above)
Comment (bhavyadeep-purswani) [reply Assignment1/assignment1/test/assignment1_test.dart:42]: Figure out a way to add test cases for generateBill with positive and negative flows

File: Assignment1/assignment1/lib/assignment1.dart (hunk 3 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/assignment1.dart]: static const String RAW_ITEM_TYPE = 'raw';

File: Assignment1/assignment1/lib/Types.dart (hunk 10 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: static const IMPORT_DUTY_PERCENTAGE = 0.1;

File: Assignment1/assignment1/lib/Types.dart (hunk 1 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: Data member - tax

File: Assignment1/assignment1/test/assignment1_test.dart (hunk 4 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/test/assignment1_test.dart:42]: for wrong item type

File: Assignment1/assignment1/lib/Types.dart (hunk 1 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: done

File: Assignment1/assignment1/lib/Types.dart (hunk 2 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: done

File: Assignment1/assignment1/lib/assignment1.dart (hunk 3 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/assignment1.dart]: done

File: Assignment1/assignment1/lib/Types.dart (hunk 5 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: done

File: Assignment1/assignment1/lib/Types.dart (hunk 6 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: done

File: Assignment1/assignment1/lib/handler.dart (hunk 9 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/handler.dart]: done

File: Assignment1/assignment1/test/assignment1_test.dart (hunk 4 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/test/assignment1_test.dart:42]: done

File: Assignment1/assignment1/lib/Types.dart (hunk 10 above)
Comment (Bhavishya-T) [reply Assignment1/assignment1/lib/Types.dart]: done
